sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V1FallbackWriters.scala

```diff
@@ -0,0 +1,115 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.SparkException
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{Dataset, SaveMode}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.sources.{AlwaysTrue, CreatableRelationProvider, Filter}
+import org.apache.spark.sql.sources.v2.Table
+import org.apache.spark.sql.sources.v2.writer._
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+/**
+ * Physical plan node for append into a v2 table using V1 write interfaces.
+ *
+ * Rows in the output data set are appended.
+ */
+case class AppendDataExecV1(
+    writeBuilder: V1WriteBuilder,
+    writeOptions: CaseInsensitiveStringMap,
+    plan: LogicalPlan) extends V1FallbackWriters {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    writeWithV1(writeBuilder.buildForV1Write(), writeOptions)
+  }
+}
+
+/**
+ * Physical plan node for overwrite into a v2 table with V1 write interfaces. Note that when this
+ * interface is used, the atomicity of the operation depends solely on the target data source.
+ *
+ * Overwrites data in a table matched by a set of filters. Rows matching all of the filters will be
+ * deleted and rows in the output data set are appended.
+ *
+ * This plan is used to implement SaveMode.Overwrite. The behavior of SaveMode.Overwrite is to
+ * truncate the table -- delete all rows -- and append the output data set. This uses the filter
+ * AlwaysTrue to delete all rows.
+ */
+case class OverwriteByExpressionExecV1(
+    table: Table,
+    writeBuilder: V1WriteBuilder,
+    deleteWhere: Array[Filter],
+    writeOptions: CaseInsensitiveStringMap,
+    plan: LogicalPlan) extends V1FallbackWriters {
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
+  }
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    writeBuilder match {
+      case builder: SupportsTruncate if isTruncate(deleteWhere) =>
+        writeWithV1(builder.truncate().asV1Writer.buildForV1Write(), writeOptions)
```

Comment:
Then it's not a simple fallback now. People need to implement `SupportsTruncate` and `SupportsOverwrite`, and people need to update their `CreatableRelationProvider` implementation to apply the `deleteWhere` condition.

AFAIK a v1 source can write data with SaveMode, which has inconsistent behavior and we don't want to rely on it. A v1 source can also write data with `InsertableRelation.insert`, which can append data or overwrite the entire table. I think those are the only 2 cases we can safely fallback to v1 source.

Reply:
I've thought about it more, and since the only mode of write that we want to support is Appending, I changed the behavior to `InsertableRelation`. With regards to:

> Then it's not a simple fallback now. People need to implement SupportsTruncate and SupportsOverwrite.

I think this is fine. If a data source already has implemented SaveMode.Overwrite, they already have some logic to delete partitions. Implementing these as a part of the WriteBuilder API shouldn't be too hard, and code re-use is possible.

Reply:
An alternative is `SupportsDelete`, which could be used for both. We just can't support dynamic overwrite.